rsocket/internal/ManageableConnection.h

```diff
@@ -0,0 +1,53 @@
+// Copyright 2004-present Facebook. All Rights Reserved.
+#pragma once
+
+#include <folly/ExceptionWrapper.h>
+#include <folly/futures/SharedPromise.h>
+#include <folly/io/async/EventBase.h>
+
+namespace rsocket {
+
+class MockManageableConnection;
+enum class StreamCompletionSignal;
+
+/// Utility class that enables RSocketConnectionManager to handle
+/// lifetime of RSocketStateMachine objects.
+class ManageableConnection {
+public:
+  virtual ~ManageableConnection() {
+    if (!closePromise_.isFulfilled()) {
+      // don't let any object indefinitely wait for the close event.
+      closePromise_.setValue();
+    }
+  }
+
+  /// Listen for the close event.
+  ///
+  /// The proposed usage is utilizing `via` function and appending
+  /// more functionality with `then` function.
+  ///
+  /// \return a Future to bind to.
+  folly::Future<folly::Unit> listenCloseEvent() noexcept {
+    return closePromise_.getFuture();
+  }
+
+  /// Terminates underlying connection.
+  ///
+  /// This may synchronously deliver terminal signals to all
+  /// StreamAutomatonBase attached to this ConnectionAutomaton.
+  virtual void close(folly::exception_wrapper,
+                     StreamCompletionSignal) noexcept = 0;
+
+protected:
+  void onClose(folly::exception_wrapper) noexcept {
+    // ignore the exception
+    closePromise_.setValue();
+  }
+
+private:
+  folly::SharedPromise<folly::Unit> closePromise_;
```

Comment:
Why a `SharedPromise`?  Don't we expect there to be a single call of `RSocketConnectionManager::manageConnection()` per connection lifetime?

Also doesn't this promise need to be reset to a new one after closing?  What happens if we do SETUP -> CLOSE -> RESUME -> CLOSE?  We'll set the same promise twice, right?

Reply:
I thought CLOSE is terminal, I saw that comment in the RSocketConnectionEvents interface. Is this comment wrong?

  // This gets called when the RSocketStateMachine is closed.  You cant use this
  // RSocketStateMachine anymore.

Reply:
I thought we might have multiple listeners for the close event. Namely, now the connection manager listens for the close event, but it does not mean that we might also have one more listener for that event. So all of them should be able to listen that event, that's the reason of having it a SharedPromise.

Check the unit tests too, you will notice that in there we listen if the connection is closed, while the ConnectionManager is also listening for that.

I will push the Promise to the RSocketStateMachine and will make the statemachine to reset it if RESUME or SETUP is called.

Reply:
Wait, you're right, `SETUP -> CLOSE -> RESUME -> CLOSE` can't happen.  I keep confusing `CLOSE` with `DISCONNECT`...